File: java/S947.java
import java.util.HashMap;
import java.util.Map;

class UnionFind {

    private Map<Integer, Integer> parent;
    private int count;

    public UnionFind() {
        this.parent = new HashMap<>();
        this.count = 0;
    }

    public int getCount() {
        return count;
    }

    public int find(int x) {
        if (!parent.containsKey(x)) {
            parent.put(x, x);
            // 并查集集中新加入一个结点，结点的父亲结点是它自己，所以连通分量的总数 +1
            count++;
        }

        if (x != parent.get(x)) {
            parent.put(x, find(parent.get(x)));
        }
        return parent.get(x);
    }

    public void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) {
            return;
        }

        parent.put(rootX, rootY);
        // 两个连通分量合并成为一个，连通分量的总数 -1
        count--;
    }
}

class Solution_947 {

    public int removeStones(int[][] stones) {
        UnionFind unionFind = new UnionFind();

        for (int[] stone : stones) {
            // 下面这三种写法任选其一
            unionFind.union(~stone[0], stone[1]);
            // unionFind.union(stone[0] - 10001, stone[1]);
//            unionFind.union(stone[0] + 10001, stone[1]);
        }
        return stones.length - unionFind.getCount();
    }
}

public class S947 {
    public static void main(String[] args) {
        System.out.println(new Solution_947().removeStones(new int[][]{{0, 0}, {0, 1}, {1, 0}, {1, 2}, {2, 1}, {2, 2}}));
    }
}

File: python/union_set/s947.py
from typing import List


class UnionSet():
    def __init__(self):
        self.parent = {}
        self.count = 0

    def find(self, x):
        if x not in self.parent:
            self.parent[x] = x
            self.count += 1
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a, b):
        pa = self.find(a)
        pb = self.find(b)
        if pa == pb:
            return
        self.count -= 1
        self.parent[pa] = pb


class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        uniset = UnionSet()
        for x, y in stones:
            uniset.union(10001 - x, y)
        return len(stones) - uniset.count
